src/mongod_helpers.py

```diff
@@ -132,9 +133,94 @@ def initialise_replica_set(self, hosts: list) -> None:
             raise e
+        except OperationFailure as e:
+            logger.error(
+                "cannot initialise replica set: database operation failed: error: %s", str(e)
+            )
+            raise e
         finally:
             client.close()
 
+    def reconfigure_replica_set(self) -> None:
+        """Reconfigure the MongoDB replica set."""
+        replica_set_client = self.client()
+        try:
+            # get current configuration and update with correct hosts
+            rs_config = replica_set_client.admin.command("replSetGetConfig")
+            rs_config["config"]["_id"] = self._replica_set_name
+            rs_config["config"]["members"] = self.replica_set_config(rs_config)
+            rs_config["config"]["version"] += 1
+            replica_set_client.admin.command("replSetReconfig", rs_config["config"])
+        except ConnectionFailure as e:
+            logger.error(
+                "cannot reconfigure replica set: failure to connect to mongo client: error: %s",
+                str(e),
+            )
+            raise e
+        except ConfigurationError as e:
+            logger.error(
+                "cannot reconfigure replica set: incorrect credentials: error: %s", str(e)
+            )
+            raise e
+        except OperationFailure as e:
+            logger.error(
+                "cannot reconfigure replica set: database operation failed: error: %s", str(e)
+            )
+            raise e
+        finally:
+            replica_set_client.close()
+
+    def replica_set_config(self, rs_config) -> List[dict]:
+        """Maps unit ips to MongoDB replica set ids.
+
+        When reconfiguring a replica set it is a requirement of mongod that machines do not get
+        reassigned to different replica ids. This function ensures that already assigned machines
+        maintain their replica set id and that new machines get assigned a new unused id.
+
+        Args:
+            rs_config: dict containing current configuration of replica set according to mongod
+
+        Returns:
+            A list of dicts, where each dict contains the host machine with its replica set id
+        """
+        new_config = []
+
+        # acquire current members in config
+        current_members = rs_config["config"]["members"]
+        logger.debug("current config for replica set: %s", current_members)
+
+        # look at each member in the current config and decide if it belongs in the new confg
+        used_ids = set()
+        assigned_ips = set()
+        for member in current_members:
+            # get member ip without ":PORT"
+            member_ip = member["host"].split(":")[0]
+
+            # if this ip is still being used retain it in new config
+            if member_ip in self._unit_ips:
+                new_config.append(member)
+                used_ids.add(int(member["_id"]))
+                assigned_ips.add(member_ip)
```

Comment:
Not mandatory, but could be similar to:
```python
used_ids = {member["_id"] for member in current_members if member_ip in self._unit_ips}
```

Reply:
In my first approach I did this, but it started to look messy when I was trimming off the port. But why I decided not to do this was because I needed to do some other things when seeing `if member_ip in self._unit_ips` ie
1. update new config 
2.update which ips are assigned.

Reply:
Yes, a bit more efficient without multiple list comprehensions, agreed.